validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java

```diff
@@ -380,4 +393,103 @@ private BLSSignature getSignature(Bytes message, Bytes domain, BLSPublicKey sign
         MessageSignerGrpc.newBlockingStub(validators.get(signer).getChannel()).signMessage(request);
     return BLSSignature.fromBytes(Bytes.wrap(response.getMessage().toByteArray()));
   }
+
+  public Eth1Data get_eth1_vote(BeaconState state, UnsignedLong previous_eth1_distance) {
+    processEth1DataCache();
+    List<Eth1Data> new_eth1_data =
+        LongStream.range(
+                ETH1_FOLLOW_DISTANCE.longValue(),
+                ETH1_FOLLOW_DISTANCE.times(UnsignedLong.valueOf(2)).longValue())
+            .mapToObj(value -> get_eth1_data(UnsignedLong.valueOf(value)))
+            .collect(Collectors.toList());
+    List<Eth1Data> all_eth1_data =
+        LongStream.range(ETH1_FOLLOW_DISTANCE.longValue(), previous_eth1_distance.longValue())
+            .mapToObj(value -> get_eth1_data(UnsignedLong.valueOf(value)))
+            .collect(Collectors.toList());
+
+    List<Eth1Data> valid_votes = new ArrayList<Eth1Data>();
+
+    ListIterator<Eth1Data> eth1_data_votes = state.getEth1_data_votes().listIterator();
+    while (eth1_data_votes.hasNext()) {
+      UnsignedLong slot = UnsignedLong.valueOf(eth1_data_votes.nextIndex());
+      Eth1Data vote = eth1_data_votes.next();
+
+      boolean period_tail =
+          slot.mod(UnsignedLong.valueOf(SLOTS_PER_ETH1_VOTING_PERIOD))
+                  .compareTo(
+                      UnsignedLong.valueOf(
+                          (int) Math.floor(Math.sqrt(SLOTS_PER_ETH1_VOTING_PERIOD))))
+              >= 0;
+      if (new_eth1_data.contains(vote) || (period_tail && all_eth1_data.contains(vote)))
+        valid_votes.add(vote);
+    }
+
+    Eth1Data max =
+        valid_votes.stream()
+            .map(
+                vote ->
+                    new MutablePair<>(
+                        Collections.frequency(valid_votes, vote) - all_eth1_data.indexOf(vote),
+                        vote))
+            .max(
+                new Comparator<MutablePair<Integer, Eth1Data>>() {
+                  @Override
+                  public int compare(
+                      MutablePair<Integer, Eth1Data> o1, MutablePair<Integer, Eth1Data> o2) {
+                    return o1.getLeft().compareTo(o2.getLeft());
+                  }
+                })
+            .orElse(new MutablePair<>())
+            .getRight();
+    return (max != null) ? max : get_eth1_data(ETH1_FOLLOW_DISTANCE);
```

Comment:
This can be simplified to:
```
return valid_votes.stream()
        .map(
            vote ->
                new MutablePair<>(
                    Collections.frequency(valid_votes, vote) - all_eth1_data.indexOf(vote), vote))
        .max(Comparator.comparing(MutablePair::getLeft))
        .map(Pair::getRight)
        .orElseGet(() -> get_eth1_data(ETH1_FOLLOW_DISTANCE));
```

Reply:
Pure magic. You've studied the dark arts.

Reply:
Fixed